Test Case Software Update 2.5

jtrejoespinoza-grid opened the pull request:
#### Summary

This test case automates the Test Case SU 2.5.
Verify the DUT behaves according the spec  when is applying the software update.
In this test case Im using event handlers to verify the updates are happening in the expected timeframes, for this case Im validating updates are not happening and start in the limits of the expected values.

#### Related issues

Fixes: https://github.com/project-chip/matter-test-scripts/issues/656

#### Testing

This test require to restart the requestor app on each step or restore the device to previous software (restore) on each step as the we are only using only one image to update the device.

**To test locally simulating the CI :**

`python3 scripts/tests/run_python_test.py --load-from-env /tmp/test_env.yaml --script src/python_testing/TC_SU_2_5.py`

> /tmp/test_env.yaml

```
OTA_PROVIDER_APP: out/debug/chip-ota-provider-app
OTA_REQUESTOR_APP: out/debug/chip-ota-requestor-app
SU_OTA_REQUESTOR_V2: objdir-clone/chip-ota-requestor-app_v2.min.ota
TRACE_TEST_PERFETTO: out/trace_data/test
TRACE_TEST_JSON: out/trace_data/test-
TRACE_APP: out/trace_data/app-

```

**To test this as Real Device,  run  as follow:**

> 1st Terminal

`./out/debug/chip-ota-requestor-app --discriminator 123 --passcode 2123   --KVS /tmp/chip_kvs_requestor --autoApplyImage`

Launch the test case
> 2nd Terminal

`src/python_testing/TC_SU_2_5.py  --commissioning-method on-network --discriminator 123 --passcode 2123 --endpoint 0  --string-arg provider_app_path:out/debug/chip-ota-provider-app --string-arg ota_image:objdir-clone/chip-ota-requestor-app_v2.min.ota --int-arg ota_image_expected_version:2 --int-arg ota_provider_port:5541 --int-arg ota_image_download_timeout:300 `

The test case will start runnig and in the end of  each Test step you will notice in terminal with test running a message as follow:

```
[MatterTest] 12-11 15:13:27.606 INFO ========= USER PROMPT for Endpoint 0 =========
[MatterTest] 12-11 15:13:27.607 INFO >>> Manually restore the DUT to it's original version. Please type Enter when its ready. (press enter to confirm)
```
When this message show up you can kill and relaunch the requestor process by just Ctrl+C and relaunching the previous command or use the same command for the requestor, then Press Enter in the terminal running the Software Update test case.
> Previous Command

`./out/debug/chip-ota-requestor-app --discriminator 123 --passcode 2123   --KVS /tmp/chip_kvs_requestor --autoApplyImage
`

Perform following actions on each step until complete.

**On real devices the device must be restore to its original version, this mean if the the device was updated to version 2 it should be restored to version 1  in order to continue the testing.**

jtrejoespinoza-grid committed:
SU 2.5 Step 1 and 2

jtrejoespinoza-grid committed:
Added basic logic for step 3

jtrejoespinoza-grid committed:
Added logic for step 4

jtrejoespinoza-grid committed:
Completed step 5 ( still need to check if we can confirm the downloaded image was deleted

jtrejoespinoza-grid committed:
Added step 5 , verify downloaded file check

jtrejoespinoza-grid committed:
Changes  for SUBase

jtrejoespinoza-grid committed:
Merge branch 'project-chip:master' into tc_softwareupdate2_5

jtrejoespinoza-grid committed:
Incomming changes from su_base_class branch

jtrejoespinoza-grid committed:
Added line to show when the wait time of 3 minutes has completed.

jtrejoespinoza-grid committed:
Updated wait times as the spec. Added secondary implementnation for event tracking as comments.

jtrejoespinoza-grid committed:
Ruff and isort fix

jtrejoespinoza-grid committed:
Added better reading for step2: Added attribute wait for BasicInformationCluster.Softwareversion to  update it and verify it did not happend in less than the delaytime

jtrejoespinoza-grid committed:
isort check

jtrejoespinoza-grid committed:
Merge branch 'master' into tc_softwareupdate2_5

jtrejoespinoza-grid committed:
Revert "Incomming changes from su_base_class branch"

This reverts commit d43bfcaf1c429e65baf164e4c7e7bab9a1fde48a.

jtrejoespinoza-grid committed:
Merge branch 'master' into tc_softwareupdate2_5

jtrejoespinoza-grid committed:
base class changes

jtrejoespinoza-grid committed:
Updates for 2.5

jtrejoespinoza-grid committed:
Interval fix

jtrejoespinoza-grid committed:
New updates for handling await time after apply or delayapplyupdate

jtrejoespinoza-grid committed:
Fix invalid method for event handler

jtrejoespinoza-grid committed:
Updated to user test arguments

jtrejoespinoza-grid committed:
Revert "base class changes"

This reverts commit 56823c7d09c5d886b53339e0c4a003df93f3dfef.

jtrejoespinoza-grid committed:
Merge branch 'master' into tc_softwareupdate2_5

jtrejoespinoza-grid committed:
Integrated su_base_class branch

jtrejoespinoza-grid committed:
Updated test to use test arguments

jtrejoespinoza-grid committed:
Remove read_from_logs. Verify with update state idle

jtrejoespinoza-grid committed:
Revert "Integrated su_base_class branch"

This reverts commit e044f1c0e9a243a5c3170ea56b455da65740c1b3.

jtrejoespinoza-grid committed:
Merge branch 'master' into tc_softwareupdate2_5

jtrejoespinoza-grid committed:
Merge base class: Remove if needed

jtrejoespinoza-grid committed:
ruff Changes

jtrejoespinoza-grid committed:
Merge branch 'master' into tc_softwareupdate2_5

jtrejoespinoza-grid committed:
Minor changes for SU 2.5. Added restart and state value tracking.

jtrejoespinoza-grid committed:
Stop the app before start. Needed to avoid port collissions for apps launched with --secure-device-port. Removed break in monitor_app_requestor to allow multiple restarts

jtrejoespinoza-grid committed:
Added terminate method

jtrejoespinoza-grid committed:
Merge branch 'master' into tc_softwareupdate2_5

jtrejoespinoza-grid committed:
Updated startup timeout for provideradd

jtrejoespinoza-grid committed:
Port changed

jtrejoespinoza-grid committed:
Updated initialization string

jtrejoespinoza-grid committed:
Merge branch 'master' into tc_softwareupdate2_5

jtrejoespinoza-grid committed:
Removed default port for requestor.

jtrejoespinoza-grid commented on src/python_testing/TC_SU_2_5.py:
```diff
@@ -0,0 +1,507 @@
+        self.ota_image = self.user_params.get('ota_image', None)
+        self.expected_software_version = self.user_params.get('ota_image_expected_version')
+        self.provider_app_path = self.user_params.get('provider_app_path', None)
+        self.ota_provider_port = self.user_params.get('ota_provider_port', 5541)
```
These possible  errors are catched in the provider_start method with proper error logging.

jtrejoespinoza-grid committed:
Added argumetn validation

jtrejoespinoza-grid committed:
Awaiting async method

jtrejoespinoza-grid committed:
Merge branch 'master' into tc_softwareupdate2_5

jtrejoespinoza-grid committed:
Added optional values for provider log and kvs

jtrejoespinoza-grid committed:
Merge branch 'master' into tc_softwareupdate2_5

jtrejoespinoza-grid committed:
Removed TC_SU_2_5 from CI as it takes a lot of time.

restyled-commits committed:
Restyled by prettier-yaml

juandediosg commented on src/python_testing/TC_SU_2_5.py:
```diff
@@ -0,0 +1,514 @@
+        update_state_match = AttributeMatcher.from_callable(
+            "Update state is Applying",
+            lambda report: report.value == Clusters.OtaSoftwareUpdateRequestor.Enums.UpdateStateEnum.kApplying)
+        update_state_attr_handler.await_all_expected_report_matches([update_state_match], timeout_sec=60*5)
```
Same here, suggest `await` this call, since `start() `and `cancel() `are already awaited,

juandediosg commented on src/python_testing/TC_SU_2_5.py:
```diff
@@ -0,0 +1,514 @@
+        await self.verify_version_applied_basic_information(controller=self.controller, node_id=self.requestor_node_id, target_version=self.expected_software_version)
+        self.terminate_provider()
+        self.restart_requestor(self.controller)
+
```
Same case here in Step 3:

- `await_all_expected_report_matches` is not awaited; consider using `await` to keep the async flow consistent.
- Fixed `timeouts` are used for `UpdateState` transitions (`Downloading` and `DelayedOnApply`); dynamic tracking or combined matchers would make this more robust.
- Waiting for `SoftwareVersion` relies on a fixed `spec_wait_time`; according to the spec, DUT may delay applying updates longer

jtrejoespinoza-grid commented on src/python_testing/TC_SU_2_5.py:
```diff
@@ -0,0 +1,514 @@
+        update_state_match = AttributeMatcher.from_callable(
+            "Update state is Downloading",
+            lambda report: report.value == Clusters.OtaSoftwareUpdateRequestor.Enums.UpdateStateEnum.kDownloading)
+        update_state_attr_handler.await_all_expected_report_matches([update_state_match], timeout_sec=60)
```
Do you mean the call to method  `await_all_expected_report_matches` ?

jtrejoespinoza-grid committed:
Added custom download time for ota image. Check on download state before waiting the kDelayedOnApply

jtrejoespinoza-grid commented on src/python_testing/TC_SU_2_5.py:
```diff
@@ -0,0 +1,514 @@
+    provider_node_id = 321
+    provider_discriminator = 321
+    provider_setup_pincode = 2321
+
```
These are just definitions for the class.

juandediosg reviewed (approved):
All good now, approved

jtrejoespinoza-grid committed:
Added remove provider kvs at teardown

jtrejoespinoza-grid committed:
Added clear_kvs for provider

jtrejoespinoza-grid committed:
Merge branch 'master' into tc_softwareupdate2_5

jtrejoespinoza-grid committed:
Ruff check changes

jtrejoespinoza-grid committed:
Clear kvs provider

jtrejoespinoza-grid committed:
Revert file added

cecille commented on src/python_testing/TC_SU_2_5.py:
```diff
@@ -0,0 +1,509 @@
+            f"Waiting Sofware version update to {self.expected_software_version}",
+            lambda report: report.value > current_sw_version)
+        software_version_attr_handler.await_all_expected_report_matches(
+            [update_sw_vesion], timeout_sec=(delayed_apply_action_time+10))
```
same below.

cecille commented on src/python_testing/TC_SU_2_5.py:
```diff
@@ -0,0 +1,509 @@
+        asserts.assert_greater(ota_file_data['size'], 0, "Downloaded file is still at 0")
+
+        # Applying
+        update_state_match = AttributeMatcher.from_callable(
```
Isn't the DUT told to discontinue the update? Why would it still hit the applying state?

cecille commented on src/python_testing/TC_SU_2_5.py:
```diff
@@ -0,0 +1,509 @@
+            Clusters.OtaSoftwareUpdateRequestor, Clusters.OtaSoftwareUpdateRequestor.Attributes.UpdateState, self.controller, self.requestor_node_id, 0)
+        asserts.assert_equal(update_state, Clusters.OtaSoftwareUpdateRequestor.Enums.UpdateStateEnum.kIdle,
+                             "Update state is not idle")
+        ota_file_data = self.get_downloaded_ota_image_info()
```
This was checked above...does this change? is the download guaranteed to be complete before the earlier check?

jtrejoespinoza-grid committed:
Merge branch 'master' into tc_softwareupdate2_5

jtrejoespinoza-grid committed:
Merge branch 'master' into tc_softwareupdate2_5

jtrejoespinoza-grid committed:
Merge branch 'master' into tc_softwareupdate2_5

jtrejoespinoza-grid committed:
Added await_all_period_expected_report_matches to verify the attribute does not change in the period of time.

jtrejoespinoza-grid committed:
Implemented await_all_period_expected_report_matches into 2.5. Wait sofware update is not applied then verify if the software update was applied in the device.

jtrejoespinoza-grid committed:
cancel fix after merge from master

jtrejoespinoza-grid committed:
Fix ruff check for logging to log

jtrejoespinoza-grid committed:
Fix for ruff check at logging to log

jtrejoespinoza-grid commented:
Updates to this PR:

- Added a method on `AttributeSubscriptionHandler` called `await_all_period_expected_report_matches` which waits a period of time on an attribute on an specific value to not change. This will help in cases we need to validate an attribute did not change during a period of time.
- Removed Listen on ShutdownEvent
- Increased timeout after device updates the software version.

jtrejoespinoza-grid committed:
Added try catch to avoid the CancelledError on await after cancel subscription.

jtrejoespinoza-grid committed:
Catching CancelledError

jtrejoespinoza-grid committed:
Revert 4997533dea1669484d502f919f4161fe4a6a649f and 4997533dea1669484d502f919f4161fe4a6a649f and add --timeout to test run.

jtrejoespinoza-grid committed:
Revert "Revert 4997533dea1669484d502f919f4161fe4a6a649f and 4997533dea1669484d502f919f4161fe4a6a649f and add --timeout to test run."

This reverts commit 356e40fea801e294c019040e6919f0007367f57c.

jtrejoespinoza-grid committed:
Success message updated

jtrejoespinoza-grid committed:
Updated requestor_restart prompt

jtrejoespinoza-grid committed:
Merge branch 'master' into tc_softwareupdate2_5

jtrejoespinoza-grid committed:
Updated timeoutto 0

jtrejoespinoza-grid committed:
Addressing clear_kvs issues with /private/tmp

jtrejoespinoza-grid committed:
Merge branch 'master' into tc_softwareupdate2_5

jtrejoespinoza-grid committed:
Removed sleep after cancel. Added default timeout and flag to add the timeout to this test as is required. Removed sleep at restart_app

jtrejoespinoza-grid committed:
Added app ready pattern, restored sleep in restart to avois possible connection issues.

jtrejoespinoza-grid committed:
Added manual step in cases we need devices to apply the software update by using UserPrompts

r-recchia committed:
Merge branch 'master' into pr/jtrejoespinoza-grid/41182

r-recchia committed:
Update imports

r-recchia committed:
Remove yaml test

r-recchia committed:
Merge branch 'master' into tc_softwareupdate2_5

r-recchia committed:
fix imports

r-recchia committed:
Change timeouts

r-recchia committed:
Merge branch 'master' into tc_softwareupdate2_5

cecille commented on src/python_testing/TC_SU_2_5.py:
```diff
@@ -0,0 +1,508 @@
+        self.provider_kvs_path = self.user_params.get('provider_kvs_path', '/tmp/chip_kvs_provider')
+        self.provider_log = self.user_params.get('provider_log_path', '/tmp/provider_log_2_5.log')
+        # On average the ota image build for the CI is 1.8 MB which takes 4-5 min to download. Adjust if needed.
+        self.ota_image_download_timeout = self.user_params.get('ota_image_download_timeout', 60*5)
```
Can you add a follow up here to attempt to get rid of this parameter?

It will require some thought, but I think there's ways to tell if the download is progressing and whether we think it's stalled. I want to try to avoid a situation where the download is progressing fine, the tester didn't set the parameter quite right, and the test fails out after a frustratingly long time for no reason.

cecille commented on src/python_testing/TC_SU_2_5.py:
```diff
@@ -0,0 +1,508 @@
+        self.ota_image_download_timeout = self.user_params.get('ota_image_download_timeout', 60*5)
+        logger.info(f"Image download timeout is set to {self.ota_image_download_timeout} seconds")
+
+        if not self.provider_kvs_path.startswith('/tmp'):
```
why?

cecille replied:
Does that also work on the PI? I suppose in the end this isn't a huge deal. There's a default anyway, so I suspect that most users won't bother to override this.

jtrejoespinoza-grid committed:
Typo fix

jtrejoespinoza-grid committed:
Merge branch 'master' into tc_softwareupdate2_5

jtrejoespinoza-grid committed:
Moved image download check to CI

cecille commented on src/python_testing/matter_testing_infrastructure/matter/testing/event_attribute_reporting.py:
```diff
@@ -304,6 +304,47 @@ def wait_for_attribute_report(self, timeout_sec: float = 10):
         asserts.assert_equal(item.attribute, self._expected_attribute,
                              f"[AttributeSubscriptionHandler] Received incorrect report. Expected: {self._expected_attribute}, received: {item.attribute}")
 
+    def await_all_period_expected_report_matches(self, expected_matchers: Iterable[AttributeMatcher], timeout_sec: float = 1.0):
```
ah, ok. Would the name "ensure_no_matching_reports_for_time" or something like that be more descriptive? Also, can you please add a comment indicating when people should use this vs. the other function?

jtrejoespinoza-grid committed:
Fix: Updated method name to wait_all_final_values_reported_persisted and removed wait for kApply on Step5, instead wait for DownloadComplete

jtrejoespinoza-grid committed:
fix: ruff changes

jtrejoespinoza-grid committed:
Revert "Remove yaml test"

This reverts commit ad275f31753d2e17d8521a4b79ae28b816e7296e.

jtrejoespinoza-grid committed:
Restored yaml test for 2.7

jtrejoespinoza-grid commented:
Yaml test `Test_TC_SU_2_5.yaml` and `Test_TC_S_2_7.yaml` restored as requested.

jtrejoespinoza-grid committed:
Merge branch 'master' into tc_softwareupdate2_5
